creusot/src/translation/pearlite.rs

```diff
@@ -156,6 +156,38 @@ pub enum TermKind<'tcx> {
     },
     Absurd,
 }
+
+impl<'tcx> TermKind<'tcx> {
+    pub fn item(
+        def_id: DefId,
+        subst: GenericArgsRef<'tcx>,
+        user_ty: &Option<Box<CanonicalUserType<'tcx>>>,
+        tcx: TyCtxt<'tcx>,
+    ) -> Self {
+        let Some(user_ty) = user_ty else { return Self::Item(def_id, subst) };
+
+        match user_ty.value {
+            UserType::Ty(_) => Self::Item(def_id, subst),
+            UserType::TypeOf(def_id2, u_subst) => {
+                assert_eq!(def_id, def_id2);
+                if u_subst.args.len() != subst.len() {
+                    return Self::Item(def_id, subst);
+                }
+                let subst = GenericArgs::for_item(tcx, def_id, |x, _| {
+                    let s = subst[x.index as usize];
+                    let us = u_subst.args[x.index as usize];
+                    if us.has_escaping_bound_vars() {
+                        s
+                    } else {
+                        us
+                    }
+                });
+                Self::Item(def_id, subst)
```

Comment:
Can you explain why this is needed?

Reply:
The types in THIR have all of the lifetimes erased, this is normally fine since the Prusti translation/type-checker can get the argument and return types with lifetimes and infer the rest. The issue the translation can't see lifetimes parameters that were manually instantiated by the user. This change replaces the erased lifetimes with the ones manually instantiated by the use, or leaves them erased if they weren't manually instantiated.

The translation currently only uses this information to extract the lifetime from `at::<'x>` expressions, but eventually it could also error if the instantiated lifetime of a `logic` function doesn't match what it inferred.